src/server/unit_sao.cpp

```diff
@@ -130,50 +130,97 @@ void UnitSAO::sendOutdatedData()
 	}
 }
 
-void UnitSAO::setAttachment(int parent_id, const std::string &bone, v3f position,
+void UnitSAO::setAttachment(const object_t new_parent, const std::string &bone, v3f position,
 		v3f rotation, bool force_visible)
 {
-	auto *obj = parent_id ? m_env->getActiveObject(parent_id) : nullptr;
-	if (obj) {
-		// Do checks to avoid circular references
-		// The chain of wanted parent must not refer or contain "this"
-		for (obj = obj->getParent(); obj; obj = obj->getParent()) {
-			if (obj == this) {
-				warningstream << "Mod bug: Attempted to attach object " << m_id << " to parent "
-					<< parent_id << " but former is an (in)direct parent of latter." << std::endl;
-				return;
+	const auto call_count = ++m_attachment_call_counter;
+
+	const auto check_nesting = [&] (const char *descr) -> bool {
+		if (m_attachment_call_counter == call_count)
+			return false;
+		auto d = wrappedDifference(m_attachment_call_counter, call_count, (u8)U8_MAX);
+		if (d > 30) {
```

Comment:
Does not trigger when punching the lowest node in the given test script with the following callback code:
```Lua
	on_punch = function(self, puncher, _, _, _, _)
		self.object:set_detach()
		--error("lockout")
	end,
	on_detach = function(self, parent)
		self.object:set_attach(parent, "", {x=30,y=0,z=30}, {x=45,y=45,z=0})
		print("on_detach")
	end,
```

because `check_nesting` is only run on the most recent (or 2nd most recent when no detach is necessary) function call, thus resulting in distances of 0 and 1:

```
CALL 152
     ^ call_count
152 DETACH
^ call_count
CALL 153
153 ATTACH
	check 153 152
          ^ m_attachment_call_counter | call_count
CALL 154
154 DETACH
CALL 155
155 ATTACH
	check 155 154
CALL 156
156 DETACH
CALL 157
157 ATTACH
	check 157 156
CALL 158
158 DETACH
CALL 159
159 ATTACH
```

Suggestion: use `m_attachment_call_counter--;`  on function exit, similar to `Inventory::ResizeUnlocker`. Error when `call_counter == 255).

Reply:
I need to be able to abort an existing attach operation if a nested call to `setAttachment` has completed. this doesn't work if the value is decremented again.

you're correct that the loop detection doesn't work like this, I'll have to remove it.